crash when summon Mirror for case class with private ctor compiled by 2.13

## Compiler version

3.1.1

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```Scala
// scalac -version ==> 2.13.6
case class Foo private (i: Int)
```

```Scala
// scalac -version ==> 3.1.1
@main def Test =
  summon[deriving.Mirror.Of[Foo]]
```


## Output (click arrow to expand)
<details>

```scala
java.lang.AssertionError: assertion failed: private constructor Foo in class Foo in <no file> accessed from method fromProduct in baz.scala while compiling baz.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: private constructor Foo in class Foo in <no file> accessed from method fromProduct in baz.scala
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.dotc.transform.ExpandPrivate.ensurePrivateAccessible(ExpandPrivate.scala:87)
	at dotty.tools.dotc.transform.ExpandPrivate.transformSelect(ExpandPrivate.scala:97)
	at dotty.tools.dotc.transform.ExpandPrivate.transformSelect(ExpandPrivate.scala:96)
	at dotty.tools.dotc.transform.MegaPhase.goSelect(MegaPhase.scala:599)
	at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:229)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
	at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:279)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:429)
	at dotty.tools.dotc.transform.MegaPhase.mapDefDef$1(MegaPhase.scala:249)
	at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:252)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
	at dotty.tools.dotc.transform.MegaPhase.transformStat$1(MegaPhase.scala:437)
	at dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
	at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:442)
	at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:362)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:429)
	at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:256)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
	at dotty.tools.dotc.transform.MegaPhase.transformStat$1(MegaPhase.scala:437)
	at dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
	at dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
	at dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
	at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:442)
	at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:362)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:429)
	at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:256)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
	at dotty.tools.dotc.transform.MegaPhase.transformStat$1(MegaPhase.scala:437)
	at dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
	at dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
	at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:442)
	at dotty.tools.dotc.transform.MegaPhase.mapPackage$1(MegaPhase.scala:382)
	at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:385)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:429)
	at dotty.tools.dotc.transform.MegaPhase.transformUnit(MegaPhase.scala:448)
	at dotty.tools.dotc.transform.MegaPhase.run(MegaPhase.scala:460)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:308)
	at scala.collection.immutable.List.map(List.scala:246)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:309)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:261)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:272)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:280)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:289)
	at dotty.tools.dotc.Run.compileSources(Run.scala:222)
	at dotty.tools.dotc.Run.compile(Run.scala:206)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
	at dotty.tools.dotc.Driver.process(Driver.scala:199)
	at dotty.tools.dotc.Driver.process(Driver.scala:167)
	at dotty.tools.dotc.Driver.process(Driver.scala:179)
	at dotty.tools.dotc.Driver.main(Driver.scala:209)
	at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/typer/Synthesizer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Synthesizer.scala
@@ -283,6 +283,22 @@ class Synthesizer(typer: Typer)(using @constructorOnly c: Context):
       case OrType(tp1, tp2) => acceptable(tp1, cls) && acceptable(tp2, cls)
       case _                => tp.classSymbol eq cls
 
+    /** for a case class, if it will have an anonymous mirror,
+     *  check that its constructor can be accessed
+     *  from the calling scope.
+     */
+    def canAccessCtor(cls: Symbol): Boolean =
+      !genAnonyousMirror(cls) || {
+        def isAccessible(sym: Symbol): Boolean = ctx.owner.isContainedIn(sym)
+        def isSub(sym: Symbol): Boolean = ctx.owner.ownersIterator.exists(_.derivesFrom(sym))
+        val ctor = cls.primaryConstructor
+        (!ctor.isOneOf(Private | Protected) || isSub(cls)) // we cant access the ctor because we do not extend cls
+        && (!ctor.privateWithin.exists || isAccessible(ctor.privateWithin)) // check scope is compatible
+      }
+
+    def genAnonyousMirror(cls: Symbol): Boolean =
+      cls.is(Scala2x) || cls.linkedClass.is(Case)
+
     def makeProductMirror(cls: Symbol): Tree =
       val accessors = cls.caseAccessors.filterNot(_.isAllOf(PrivateLocal))
       val elemLabels = accessors.map(acc => ConstantType(Constant(acc.name.toString)))
@@ -300,7 +316,7 @@ class Synthesizer(typer: Typer)(using @constructorOnly c: Context):
           .refinedWith(tpnme.MirroredElemTypes, TypeAlias(elemsType))
           .refinedWith(tpnme.MirroredElemLabels, TypeAlias(elemsLabels))
       val mirrorRef =
-        if (cls.is(Scala2x) || cls.linkedClass.is(Case)) anonymousMirror(monoType, ExtendsProductMirror, span)
+        if (genAnonyousMirror(cls)) anonymousMirror(monoType, ExtendsProductMirror, span)
         else companionPath(mirroredType, span)
       mirrorRef.cast(mirrorType)
     end makeProductMirror
@@ -321,8 +337,13 @@ class Synthesizer(typer: Typer)(using @constructorOnly c: Context):
             modulePath.cast(mirrorType)
         else
           val cls = mirroredType.classSymbol
-          if acceptable(mirroredType, cls) && cls.isGenericProduct then makeProductMirror(cls)
-          else EmptyTree
+          if acceptable(mirroredType, cls)
+            && cls.isGenericProduct
+            && canAccessCtor(cls)
+          then
+            makeProductMirror(cls)
+          else
+            EmptyTree
   end productMirror
 
   private def sumMirror(mirroredType: Type, formal: Type, span: Span)(using Context): Tree =

--- a/sbt-test/scala2-compat/i14432/app1fail/Test.scala
+++ b/sbt-test/scala2-compat/i14432/app1fail/Test.scala
@@ -1,0 +1,3 @@
+import deriving.Mirror
+
+val mFoo = summon[Mirror.Of[Foo]] // error: `Foo.<init>(Int)` is not accessible from `<empty>`.

--- a/sbt-test/scala2-compat/i14432/app1ok/Test.scala
+++ b/sbt-test/scala2-compat/i14432/app1ok/Test.scala
@@ -1,0 +1,8 @@
+import deriving.Mirror
+
+package example {
+  val mFoo = summon[Mirror.Of[Foo]] // ok, we can access Foo's ctor from here.
+}
+
+@main def Test: Unit =
+  assert(example.mFoo.fromProduct(Some(23)) == example.Foo(23))

--- a/sbt-test/scala2-compat/i14432/app2fail/Test.scala
+++ b/sbt-test/scala2-compat/i14432/app2fail/Test.scala
@@ -1,0 +1,5 @@
+package example
+
+import deriving.Mirror
+
+val mFoo = summon[Mirror.Of[Foo]] // error: `Foo.<init>(Int)` is not accessible from any class.

--- a/sbt-test/scala2-compat/i14432/build.sbt
+++ b/sbt-test/scala2-compat/i14432/build.sbt
@@ -1,0 +1,30 @@
+val scala3Version = sys.props("plugin.scalaVersion")
+val scala2Version = sys.props("plugin.scala2Version")
+
+lazy val lib1 = project.in(file("lib1"))
+  .settings(
+    scalaVersion := scala2Version
+  )
+
+lazy val lib2 = project.in(file("lib2"))
+  .settings(
+    scalaVersion := scala2Version
+  )
+
+lazy val app1fail = project.in(file("app1fail"))
+  .dependsOn(lib1)
+  .settings(
+    scalaVersion := scala3Version
+  )
+
+lazy val app1ok = project.in(file("app1ok"))
+  .dependsOn(lib1)
+  .settings(
+    scalaVersion := scala3Version
+  )
+
+lazy val app2fail = project.in(file("app2fail"))
+  .dependsOn(lib2)
+  .settings(
+    scalaVersion := scala3Version
+  )

--- a/sbt-test/scala2-compat/i14432/lib1/Foo.scala
+++ b/sbt-test/scala2-compat/i14432/lib1/Foo.scala
@@ -1,0 +1,3 @@
+package example
+
+case class Foo private[example] (i: Int)

--- a/sbt-test/scala2-compat/i14432/lib2/Foo.scala
+++ b/sbt-test/scala2-compat/i14432/lib2/Foo.scala
@@ -1,0 +1,3 @@
+package example
+
+case class Foo private (i: Int)

--- a/sbt-test/scala2-compat/i14432/test
+++ b/sbt-test/scala2-compat/i14432/test
@@ -1,0 +1,5 @@
+> lib1/compile
+> lib2/compile
+-> app1fail/compile
+> app1ok/run
+-> app2fail/compile

--- a/tests/neg/i14432.check
+++ b/tests/neg/i14432.check
@@ -1,0 +1,4 @@
+-- Error: tests/neg/i14432.scala:13:33 ---------------------------------------------------------------------------------
+13 |val mFoo = summon[Mirror.Of[Foo]] // error: no mirror found
+   |                                 ^
+   |no given instance of type deriving.Mirror.Of[example.Foo] was found for parameter x of method summon in object Predef

--- a/tests/neg/i14432.scala
+++ b/tests/neg/i14432.scala
@@ -1,0 +1,13 @@
+package example
+
+import deriving.Mirror
+
+case class Foo private (i: Int)
+
+// case object companion here prevents Foo from caching
+// the mirror in its companion, so all potential mirrors for Foo will be anonymous.
+case object Foo
+
+// however we can not provide an anonymous mirror
+// at this call site because the constructor is not accessible.
+val mFoo = summon[Mirror.Of[Foo]] // error: no mirror found

--- a/tests/neg/i14432a.check
+++ b/tests/neg/i14432a.check
@@ -1,0 +1,4 @@
+-- Error: tests/neg/i14432a.scala:14:43 --------------------------------------------------------------------------------
+14 |  val mFoo = summon[Mirror.Of[example.Foo]] // error: no mirror found
+   |                                           ^
+   |no given instance of type deriving.Mirror.Of[example.Foo] was found for parameter x of method summon in object Predef

--- a/tests/neg/i14432a.scala
+++ b/tests/neg/i14432a.scala
@@ -1,0 +1,15 @@
+import deriving.Mirror
+
+package example {
+  case class Foo private[example] (val i: Int)
+
+  // case object companion here prevents Foo from caching
+  // the mirror in its companion, so all potential mirrors for Foo will be anonymous.
+  case object Foo
+}
+
+@main def Test: Unit =
+  // however we can not provide an anonymous mirror
+  // at this call site because the constructor is not accessible.
+  val mFoo = summon[Mirror.Of[example.Foo]] // error: no mirror found
+  assert(mFoo.fromProduct(Tuple1(1)).i == 1)

--- a/tests/neg/i14432b.check
+++ b/tests/neg/i14432b.check
@@ -1,0 +1,4 @@
+-- Error: tests/neg/i14432b.scala:15:43 --------------------------------------------------------------------------------
+15 |  val mFoo = summon[Mirror.Of[example.Foo]] // error: no mirror found
+   |                                           ^
+   |no given instance of type deriving.Mirror.Of[example.Foo] was found for parameter x of method summon in object Predef

--- a/tests/neg/i14432b.scala
+++ b/tests/neg/i14432b.scala
@@ -1,0 +1,16 @@
+import deriving.Mirror
+
+package example {
+  case class Foo protected [example] (val i: Int)
+
+  // case object companion here prevents Foo from caching
+  // the mirror in its companion, so all potential mirrors for Foo will be anonymous.
+  case object Foo
+
+}
+
+class Bar extends example.Foo(23) {
+  // however we can not provide an anonymous mirror
+  // at this call site because the constructor is not accessible.
+  val mFoo = summon[Mirror.Of[example.Foo]] // error: no mirror found
+}

--- a/tests/neg/i14432c.check
+++ b/tests/neg/i14432c.check
@@ -1,0 +1,12 @@
+-- Error: tests/neg/i14432c.scala:12:18 --------------------------------------------------------------------------------
+12 |class Bar extends example.Foo(23) { // error // error: cant access private[example] ctor
+   |                  ^^^^^^^^^^^
+   |                  constructor Foo cannot be accessed as a member of example.Foo from class Bar.
+-- Error: tests/neg/i14432c.scala:12:6 ---------------------------------------------------------------------------------
+12 |class Bar extends example.Foo(23) { // error // error: cant access private[example] ctor
+   |      ^
+   |      constructor Foo cannot be accessed as a member of example.Foo from class Bar.
+-- Error: tests/neg/i14432c.scala:16:43 --------------------------------------------------------------------------------
+16 |  val mFoo = summon[Mirror.Of[example.Foo]] // error: no mirror
+   |                                           ^
+   |no given instance of type deriving.Mirror.Of[example.Foo] was found for parameter x of method summon in object Predef

--- a/tests/neg/i14432c.scala
+++ b/tests/neg/i14432c.scala
@@ -1,0 +1,18 @@
+import deriving.Mirror
+
+package example {
+  case class Foo private [example] (val i: Int)
+
+  // case object companion here prevents Foo from caching
+  // the mirror in its companion, so all potential mirrors for Foo will be anonymous.
+  case object Foo
+
+}
+
+class Bar extends example.Foo(23) { // error // error: cant access private[example] ctor
+
+  // however we can not provide an anonymous mirror
+  // at this call site because the constructor is not accessible.
+  val mFoo = summon[Mirror.Of[example.Foo]] // error: no mirror
+
+}

--- a/tests/neg/i14432d.check
+++ b/tests/neg/i14432d.check
@@ -1,0 +1,4 @@
+-- Error: tests/neg/i14432d.scala:17:45 --------------------------------------------------------------------------------
+17 |    val mFoo = summon[Mirror.Of[example.Foo]] // error
+   |                                             ^
+   |no given instance of type deriving.Mirror.Of[example.Foo] was found for parameter x of method summon in object Predef

--- a/tests/neg/i14432d.scala
+++ b/tests/neg/i14432d.scala
@@ -1,0 +1,20 @@
+import deriving.Mirror
+
+package example {
+  case class Foo protected [example] (val i: Int)
+
+  // case object companion here prevents Foo from caching
+  // the mirror in its companion, so all potential mirrors for Foo will be anonymous.
+  case object Foo
+
+}
+
+class Bar extends example.Foo(23) {
+
+  class Inner {
+    // however we can not provide an anonymous mirror
+    // at this call site because the constructor is not accessible.
+    val mFoo = summon[Mirror.Of[example.Foo]] // error
+  }
+
+}

--- a/tests/run/i14432.scala
+++ b/tests/run/i14432.scala
@@ -1,0 +1,11 @@
+import deriving.Mirror
+
+package example {
+  // Foo caches the mirror in its companion,
+  // which can still access the constructor.
+  case class Foo private (val i: Int)
+}
+
+@main def Test: Unit =
+  val mFoo = summon[Mirror.Of[example.Foo]]
+  assert(mFoo.fromProduct(Tuple1(1)).i == 1)

--- a/tests/run/i14432a.scala
+++ b/tests/run/i14432a.scala
@@ -1,0 +1,16 @@
+import deriving.Mirror
+
+package example {
+  case class Foo private[example] (val i: Int)
+
+  // case object companion here prevents Foo from caching
+  // the mirror in its companion, so all potential mirrors for Foo will be anonymous.
+  case object Foo
+
+  // here, we can synthesize an anonymous mirror
+  // because at this call site the constructor is accessible.
+  val mFoo = summon[Mirror.Of[example.Foo]]
+}
+
+@main def Test: Unit =
+  assert(example.mFoo.fromProduct(Tuple1(1)).i == 1)

--- a/tests/run/i14432b.scala
+++ b/tests/run/i14432b.scala
@@ -1,0 +1,20 @@
+import deriving.Mirror
+
+package example {
+  case class Foo protected [example] (val i: Int)
+
+  // case object companion here prevents Foo from caching
+  // the mirror in its companion, so all potential mirrors for Foo will be anonymous.
+  case object Foo
+
+  class Bar extends Foo(23) {
+    // here, we can synthesize an anonymous mirror
+    // because at this call site the constructor is accessible.
+    val mFoo = summon[Mirror.Of[example.Foo]]
+  }
+
+}
+
+@main def Test: Unit =
+  val bar = new example.Bar
+  assert(bar.mFoo.fromProduct(Tuple1(1)).i == 1)

--- a/tests/run/i14432c.scala
+++ b/tests/run/i14432c.scala
@@ -1,0 +1,23 @@
+import deriving.Mirror
+
+package example {
+  case class Foo protected [example] (val i: Int)
+
+  // case object companion here prevents Foo from caching
+  // the mirror in its companion, so all potential mirrors for Foo will be anonymous.
+  case object Foo
+
+  class Bar extends Foo(23) {
+    class Inner {
+      // here, we can synthesize an anonymous mirror
+      // because at this call site the constructor is accessible.
+      val mFoo = summon[Mirror.Of[example.Foo]]
+    }
+    val inner = Inner()
+  }
+
+}
+
+@main def Test: Unit =
+  val bar = new example.Bar
+  assert(bar.inner.mFoo.fromProduct(Tuple1(1)).i == 1)

